server/sonar-server/src/main/java/org/sonar/server/computation/step/FeedComponentsCacheStep.java

```diff
@@ -0,0 +1,109 @@
+import org.sonar.server.computation.ComputationContext;
+import org.sonar.server.computation.component.ComputeComponentsRefCache;
+import org.sonar.server.db.DbClient;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Read all components from the batch report and feed the cache containing component uuid and key
+ */
+public class FeedComponentsCacheStep implements ComputationStep {
+
+  private final DbClient dbClient;
+  private final ComputeComponentsRefCache computeComponentsRefCache;
+
+  public FeedComponentsCacheStep(DbClient dbClient, ComputeComponentsRefCache computeComponentsRefCache) {
+    this.dbClient = dbClient;
+    this.computeComponentsRefCache = computeComponentsRefCache;
+  }
+
+  @Override
+  public String[] supportedProjectQualifiers() {
+    return new String[] {Qualifiers.PROJECT};
+  }
+
+  @Override
+  public void execute(ComputationContext context) {
+    DbSession session = dbClient.openSession(false);
+    try {
+      List<ComponentDto> components = dbClient.componentDao().selectComponentsFromProjectUuid(session, context.getProject().uuid());
+      Map<String, ComponentDto> componentDtosByKey = componentDtosByKey(components);
+      int rootComponentRef = context.getReportMetadata().getRootComponentRef();
+      recursivelyProcessComponent(context, rootComponentRef, context.getReportReader().readComponent(rootComponentRef), componentDtosByKey);
+    } finally {
+      session.close();
+    }
+  }
+
+  private void recursivelyProcessComponent(ComputationContext context, int componentRef, BatchReport.Component nearestModule, Map<String, ComponentDto> componentDtosByKey) {
+    BatchReportReader reportReader = context.getReportReader();
+    BatchReport.Component reportComponent = reportReader.readComponent(componentRef);
+
+    String path = reportComponent.hasPath() ? reportComponent.getPath() : null;
+    String branch = context.getReportMetadata().hasBranch() ? context.getReportMetadata().getBranch() : null;
+    String componentKey = reportComponent.hasKey() ?
+      ComponentKeys.createKey(reportComponent.getKey(), branch) :
+      ComponentKeys.createKey(nearestModule.getKey(), path, branch);
+
+    ComponentDto componentDto = componentDtosByKey.get(componentKey);
+    if (componentDto == null) {
+      computeComponentsRefCache.addComponent(componentRef, new ComputeComponentsRefCache.ComputeComponent(componentKey, Uuids.create()));
+    } else {
+      computeComponentsRefCache.addComponent(componentRef, new ComputeComponentsRefCache.ComputeComponent(componentKey, componentDto.uuid()));
+    }
+
+    for (Integer childRef : reportComponent.getChildRefList()) {
+      // If current component is not a module or a project, we need to keep the parent reference to the nearest module
+      BatchReport.Component nextModuleParent = !reportComponent.getType().equals(Constants.ComponentType.PROJECT) && !reportComponent.getType().equals(Constants.ComponentType.MODULE) ?
```

Comment:
![New](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/new.png) ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Split this 184 characters long line (which is greater than 180 authorized). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)